flypipe/node.py

```diff
@@ -369,8 +328,49 @@ def _get_consolidated_output_schema(cls, output_schema, output_columns):
             schema = None
         return schema
 
+    def process_transformation_with_cache(
+            self, runnable_node, run_context: RunContext
+    ):
+
+        result = None
+        node_transformation = runnable_node["transformation"]
+        ran_transformation = False
+
+        if runnable_node["status"] == RunStatus.CACHED and \
+                runnable_node["node_run_context"].cache_context.exists_cache_to_load:
+            result = runnable_node["node_run_context"].cache_context.read()
+        else:
+            ran_transformation = True
+            dependency_values = runnable_node["transformation"].get_node_inputs(
+                run_context
+            )
+
+            result = node_transformation.process_transformation(
+                run_context.spark,
+                runnable_node["output_columns"],
+                runnable_node["node_run_context"],
+                **dependency_values,
+            )
+
+        run_context.update_node_results(
+            node_transformation.key,
+            result,
+            schema=self._get_consolidated_output_schema(
+                node_transformation.output_schema,
+                runnable_node["output_columns"],
+            ),
+        )
+
+        if ran_transformation:
+            # If cache exists, and the transformation has run, then save its cache
+            runnable_node["node_run_context"].cache_context.write(
+                run_context.node_results[node_transformation.key]
+                .as_type(node_transformation.dataframe_type)

```

Comment:
Can we move the casting/boxing/unboxing, etc to process_transformation if it's required? I think you should handle only caching in this function.

Reply:
cahce is dependent on the node rresult as it converts to the appropriate dataframe type and columns selection from the node output schema.

process_transformation is unaware of the output schema and dataframe conversion